File: jmlparser-jml-tools/src/test/java/com/github/jmlparser/jml2java/Jml2JavaTests.java
package com.github.jmlparser.jml2java;

import com.github.javaparser.ParseResult;
import com.github.javaparser.ast.expr.Expression;
import com.github.javaparser.printer.DefaultPrettyPrinter;
import com.github.jmlparser.TestWithJavaParser;
import com.google.common.truth.Truth;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;
import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.fail;

/**
 * @author Alexander Weigl
 * @version 1 (04.10.22)
 */
public class Jml2JavaTests extends TestWithJavaParser {
    @TestFactory
    public Stream<DynamicTest> j2jTranslation() throws IOException {
        try (InputStream inputStream = getClass().getResourceAsStream("expr.yaml")) {
            Yaml yaml = new Yaml();
            List<Map<String, Object>> obj = yaml.load(inputStream);
            return obj.stream().map(m -> {
                String a = (String) m.get("expr");
                String result = (String) m.get("result");
                return DynamicTest.dynamicTest(a, () -> {
                    if (result != null)
                        jml2JavaTranslation(a, result);
                });
            });
        }
    }

    void jml2JavaTranslation(String input, String expected) {
        ParseResult<Expression> e = parser.parseJmlExpression(input);
        if (!e.isSuccessful()) {
            e.getProblems().forEach(System.err::println);
            fail("Error during parsing");
        }
        Expression expr = e.getResult().get();
        expr.setParentNode(parent);
        var actual = Jml2JavaFacade.translate(expr);

        DefaultPrettyPrinter dpp = new DefaultPrettyPrinter();
        var sblock = dpp.print(actual.a);
        var sexpr = dpp.print(actual.b);
        Truth.assertThat(trimAllWs(sblock + " " + sexpr))
                .isEqualTo(trimAllWs(expected));
    }

    private static String trimAllWs(String expected) {
        return expected.replaceAll("\\s+", " ").trim();
    }
}

File: tools/jml2java/src/test/kotlin/io/github/jmltoolkit/jml2java/Jml2JavaTests.kt
package io.github.jmltoolkit.jml2java

import com.github.javaparser.ParseResult
import com.github.javaparser.ast.CompilationUnit
import com.github.javaparser.ast.expr.Expression
import com.github.javaparser.printer.DefaultPrettyPrinter
import com.github.javaparser.symbolsolver.JavaSymbolSolver
import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver
import com.google.common.truth.Truth
import io.github.jmltoolkit.utils.TestWithJavaParser
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import org.yaml.snakeyaml.Yaml
import java.io.IOException

data class ExprTestCase(
    val expr: String,
    val result: String,
    val disabled: Boolean = false
) {
    constructor(m: Map<String, Any?>)
            : this(
        m["expr"] as String,
        m["result"] as String,
        m["disabled"] as? Boolean ?: false
    )

}


/**
 * @author Alexander Weigl
 * @version 1 (04.10.22)
 */
class Jml2JavaTests : TestWithJavaParser() {
    val pseudoCompilationUnit = CompilationUnit().also {
        JavaSymbolSolver(ReflectionTypeSolver(true)).inject(it)
    }

    @TestFactory
    @Throws(IOException::class)
    fun j2jTranslation(): List<DynamicTest> {
        javaClass.getResourceAsStream("/expr.yaml").use { inputStream ->
            val yaml = Yaml()
            val obj: List<Map<String, Any>> = yaml.load(inputStream)
            println(obj)
            return obj.map { ExprTestCase(it) }
                .filter { !it.disabled }
                .map { DynamicTest.dynamicTest(it.expr) { jml2JavaTranslation(it) } }
        }
    }

    private fun jml2JavaTranslation(input: ExprTestCase) {
        val e: ParseResult<Expression> = parser.parseJmlExpression(input.expr)
        if (!e.isSuccessful) {
            e.problems.forEach { System.err.println(it) }
            error("Error during parsing")
        }
        val expr = e.result.get()
        expr.setParentNode(pseudoCompilationUnit)
        val actual = Jml2JavaFacade.translate(expr)

        val dpp = DefaultPrettyPrinter()
        val sblock = dpp.print(actual.a)
        val sexpr = dpp.print(actual.b)
        Truth
            .assertThat(trimAllWs("$sblock $sexpr"))
            .isEqualTo(trimAllWs(input.result))
    }
}

private fun trimAllWs(expected: String): String = expected.replace("\\s+".toRegex(), " ").trim { it <= ' ' }